src/lib/__tests__/routing/spa-debounced-test.ts

```diff
@@ -0,0 +1,93 @@
+import { createSearchClient } from '../../../../test/mock/createSearchClient';
+import { wait } from '../../../../test/utils/wait';
+import historyRouter from '../../routers/history';
+import instantsearch from '../../..';
+import { connectSearchBox } from '../../../connectors';
+
+const writeDelay = 10;
+const writeWait = 1.5 * writeDelay;
+
+describe('routing', () => {
+  beforeEach(() => {
+    window.history.replaceState({}, '', 'http://localhost/');
+    jest.clearAllMocks();
+  });
+
+  test('SPA debounced (Single Page App) use case: URL should not be cleaned', async () => {
+    // --- Flow
+    // Initial: '/'
+    // Refine: '/?indexName[query]=Apple'
+    // Dispose: '/'
+    // Route change: '/about'
+    // Back: '/'
+    // Back: '/?indexName[query]=Apple'
+
+    const pushState = jest.spyOn(window.history, 'pushState');
+
+    const search = instantsearch({
+      indexName: 'indexName',
+      searchClient: createSearchClient(),
+      routing: {
+        router: historyRouter({
+          writeDelay,
+        }),
+      },
+    });
+
+    search.addWidgets([connectSearchBox(() => {})({})]);
+
+    search.start();
+
+    // Check URL has been initialized
+    await wait(writeWait);
+    expect(window.location.search).toEqual('');
+    expect(pushState).toHaveBeenCalledTimes(0);
+
+    // Trigger an update by refining
+    search.renderState.indexName!.searchBox!.refine('Apple');
+
+    // Check URL has been updated
+    await wait(writeWait);
+    expect(window.location.search).toEqual(
+      `?${encodeURI('indexName[query]=Apple')}`
+    );
+    expect(pushState).toHaveBeenCalledTimes(1);
+
+    // Trigger a dispose
+    search.dispose();
+
+    // Check URL has been updated to new page
+    await wait(writeWait);
+    expect(window.location.pathname).toEqual('/');
+    expect(window.location.search).toEqual('');
+    expect(pushState).toHaveBeenCalledTimes(2);
+
+    // Navigate to a new page (like a router would do)
+    window.history.pushState({}, '', '/about');
```

Comment:
In our testing, this behaviour doesn't actually happen, as route navigations happen immediately, no debounce.

We are testing it to be aware if it ever changes, but that doesn't mean we think this is 100% correct behaviour (ideally the in-between update doesn't happen either.

How could that be made more clear in the test?

Reply:
The scenario as I understand it is:

- User refines InstantSearch, route updates (`/?indexName[query]=Apple`)
- User clicks navigation handled by 3rd party router, but router doesn't kick in right away because it's debounced (longer delay than ours)
- InstantSearch is disposed, history length is the same, so it cleans (`/`)
- Router finally writes (`/about`)

If yes I think this is clear, but we should probably document it because having this intermediary `/` route can be surprising.